Base/Python/slicer/packaging/requirements.py

```diff
@@ -0,0 +1,301 @@
+"""
+Classes and functions related to automatic installation of Python packages via the
+``Requirements`` context manager.
+"""
+
+import importlib.abc
+import importlib.machinery
+import importlib.resources
+import importlib.util
+import inspect
+import sys
+import types
+import typing
+import warnings
+from typing import Optional, Union
+
+import slicer
+from slicer.ScriptedLoadableModule import ScriptedLoadableModule
+
+from . import installer
+
+__all__ = ["Requirements", "LazyProxyModule", "real_module"]
+
+USE_REQUIREMENTS_AS_CONSTRAINTS = object()
+"""
+Sentinel singleton value to indicate that the ``requirements.txt`` is a valid ``constraints.txt``.
+"""
+
+
+class Requirements:
+    """
+    Encapsulates a named group of dependencies which should be resolved together using
+    ``pip_install``.
+
+    Dependencies can be resolved explicitly with ``Requirements.resolve``, or the ``Requirements``
+    object can be used as a context manager to create proxy modules which automatically call
+    ``resolve`` the first time the module is actually used.
+
+    When used as an explicit dependency group::
+
+        from slicer.packaging import Requirements
+
+        group = Requirements('package:requirements.txt')
+
+        try:
+            group.resolve()
+
+            import dependency  # Cannot fail.
+        except InstallationAbortedError:
+            ...  # User canceled installation.
+        except CalledProcessError:
+            ...  # Installation failed.
+
+    When used as a context manager::
+
+        from slicer.packaging import Requirements
+
+        with Requirements('package:requirements.txt'):
+            import dependency
+
+        dependency.function()
+        #         ^ Resolution occurs here.
+
+    See `Automatic Installation <https://slicer.readthedocs.io/en/latest/developer_guide/python_packaging.html#manual-installation>`_.
```

Comment:
This obviously doesn't work right in the CI readthedocs. I assume there is some way to get reST to do this properly?